src/pulselistener/pulselistener/mercurial.py

```diff
@@ -0,0 +1,151 @@
+
+        # Remove key when finished
+        atexit.register(self.cleanup)
+
+    def cleanup(self):
+        os.unlink(self.ssh_key_path)
+        logger.info('Removed ssh key')
+
+    async def run(self):
+        # Start by updating the repo
+        logger.info('Checking out tip', repo=self.repo_url)
+        self.repo = robust_checkout(self.repo_url, self.repo_dir)
+        self.repo.setcbout(lambda msg: logger.info('Mercurial', stdout=msg))
+        self.repo.setcberr(lambda msg: logger.info('Mercurial', stderr=msg))
+        logger.info('Initial clone finished')
+
+        # Wait for phabricator diffs to apply
+        while True:
+            diff = await self.queue.get()
+            assert isinstance(diff, dict)
+            assert 'phid' in diff
+
+            try:
+                await self.handle_diff(diff)
+
+            except hglib.error.CommandError as e:
+                logger.warn('Mercurial error on diff', error=e.err, args=e.args, phid=diff['phid'])
+
+                # Remove uncommited changes
+                self.repo.revert(self.repo_dir.encode('utf-8'), all=True)
+
+            except Exception as e:
+                logger.warn('Failed to process diff', error=e, phid=diff['phid'])
+
+                # Remove uncommited changes
+                self.repo.revert(self.repo_dir.encode('utf-8'), all=True)
+
+            # Notify the queue that the message has been processed
+            self.queue.task_done()
+
+    def clean(self):
+        '''
+        Steps to clean the mercurial repo
+        '''
+        logger.info('Remove all mercurial drafts')
+        try:
+            cmd = hglib.util.cmdbuilder(b'strip', rev=b'roots(outgoing())', force=True)
+            self.repo.rawcommand(cmd)
+        except hglib.error.CommandError as e:
+            if b'abort: empty revision set' not in e.err:
+                raise
+
+        logger.info('Pull updates from remote repo')
+        self.repo.pull()
+
+    async def handle_diff(self, diff):
+        '''
+        Handle a new diff received from Phabricator:
+        - apply revision to mercurial repo
+        - trigger push-to-try
+        '''
+        logger.info('Received diff {phid}'.format(**diff))
+
+        # Start by cleaning the repo
+        self.clean()
+
+        # Get the stack of patches
+        patches = self.phabricator_api.load_patches_stack(self.repo, diff, default_revision='central')
+        assert len(patches) > 0, 'No patches to apply'
+
+        # Get current revision
+        base = self.repo.tip()
+
+        # Apply the patches and commit them one by one
+        for diff_phid, patch in patches:
+            logger.info('Applying patch', phid=diff_phid)
+            self.repo.import_(
+                patches=io.BytesIO(patch.encode('utf-8')),
+                message='Patch {}'.format(diff_phid),
+                user='pulselistener',
+            )
+
+        # Rebase multiple patches into a single commit
```

Comment:
Squashing the patches into a single commit shouldn't be needed

Reply:
Agreed, it works without squash: https://treeherder.mozilla.org/#/jobs?repo=try&revision=c7cfff789d14b9ba7b25ad329df72639af03ba65&selectedJob=226153182